File: windows/src/CoulombMppt/Data/MpptSettings.cs
namespace CoulombMppt.Data;

// Snapshot of the controller's writable settings (BLE_PROTOCOL.md §3.2).
// Every voltage field is reg/10. Ported from the Android client's
// MpptSettings.kt.
public sealed record MpptSettings(
    int    BatteryType,              // enum code
    int    TimerHours,
    int    TimerMinutes,
    double ChargeVoltageSetpoint,    // V — "full" / boost target
    int    OutputMode,               // enum code
    double CutoffVoltageSetpoint,    // V — low-disconnect / "empty"
    bool   ManualLoadOn,
    int    VoltageMonitorMode,       // enum code
    double RecoveryVoltageSetpoint)  // V — load reconnect
{
    /// <summary>
    /// Linear-interpolation SoC from the controller's own calibration. The
    /// cutoff and charge setpoints are the controller's idea of "0%" and
    /// "100%". Returns [0,100], or null if the calibration is degenerate.
    /// </summary>
    public double? ComputeSoc(double batteryVoltage)
    {
        double low = CutoffVoltageSetpoint, high = ChargeVoltageSetpoint;
        if (high <= low + 0.1) return null;                  // degenerate calibration
        double pct = (batteryVoltage - low) / (high - low) * 100.0;
        return Math.Clamp(pct, 0.0, 100.0);
    }

    public BatteryType BatteryTypeEnum => BatteryTypeExtensions.OfCode(BatteryType);
    public OutputMode  OutputModeEnum  => OutputModeExtensions.OfCode(OutputMode);
}

// Controller's own firmware battery-type enum (codes 0-4). Distinct from
// BatteryChemistry, which is what the user told us their pack actually is.
public enum BatteryType
{
    Unknown     = 0,
    SealedLead  = 1,
    GelLead     = 2,
    FloodedLead = 3,
    Lithium     = 4,
}

public static class BatteryTypeExtensions
{
    public static BatteryType OfCode(int c) =>
        Enum.IsDefined(typeof(BatteryType), c) ? (BatteryType)c : BatteryType.Unknown;

    public static string DisplayName(this BatteryType t) => t switch
    {
        BatteryType.SealedLead  => "Sealed lead-acid",
        BatteryType.GelLead     => "Gel lead-acid",
        BatteryType.FloodedLead => "Flooded lead-acid",
        BatteryType.Lithium     => "Lithium (LiFePO4)",
        _                       => "Unknown",
    };
}

public enum OutputMode
{
    Manual  = 0,
    Auto    = 1,
    Timer   = 2,
    Unknown = 255,
}

public static class OutputModeExtensions
{
    public static OutputMode OfCode(int c) =>
        Enum.IsDefined(typeof(OutputMode), c) ? (OutputMode)c : OutputMode.Unknown;

    public static string DisplayName(this OutputMode m) => m switch
    {
        OutputMode.Manual => "Manual",
        OutputMode.Auto   => "Auto / dusk-to-dawn",
        OutputMode.Timer  => "Timer",
        _                 => "Unknown",
    };
}

File: android/app/src/main/kotlin/app/coulombmppt/data/model/MpptSettings.kt
package app.coulombmppt.data.model

import kotlinx.serialization.Serializable

// Snapshot of the controller's writable settings (BLE_PROTOCOL.md §3.2).
// `convert` is hard-coded to 10 for every voltage field, which is what the
// live telemetry uses too — confirm on hardware and tune per-field if any
// turn out to be different.
@Serializable
data class MpptSettings(
    val batteryType:               Int,      // enum — codes TBD on hardware
    val timerHours:                Int,
    val timerMinutes:              Int,
    val chargeVoltageSetpoint:     Double,   // V — reg / 10 — "full" / boost target
    val outputMode:                Int,      // enum — codes TBD
    val cutoffVoltageSetpoint:     Double,   // V — reg / 10 — low-disconnect / "empty"
    val manualLoadOn:              Boolean,  // 0 / 1
    val voltageMonitorMode:        Int,      // enum — codes TBD
    val recoveryVoltageSetpoint:   Double,   // V — reg / 10 — load reconnect
) {
    /**
     * Linear-interpolation SoC from the controller's own calibration.
     *
     * The cutoff and charge setpoints are the controller's actual idea of
     * "0%" and "100%", so we treat them as the SoC anchors. That works
     * cleanly across battery chemistries (Mark's 24 V LiFePO4: 20.0 V → 0,
     * 25.5 V → 100; a 12 V lead-acid pack: 11.1 V → 0, 14.4 V → 100).
     *
     * Returns a value in [0, 100], or null if the calibration is degenerate
     * (e.g. cutoff ≥ charge, which would happen if we read garbage from a
     * not-yet-initialised firmware).
     */
    fun computeSoc(batteryVoltage: Double): Double? {
        val low  = cutoffVoltageSetpoint
        val high = chargeVoltageSetpoint
        if (high <= low + 0.1) return null            // degenerate calibration
        val pct = (batteryVoltage - low) / (high - low) * 100.0
        return pct.coerceIn(0.0, 100.0)
    }
}

@Serializable
enum class BatteryType(val code: Int, val displayName: String) {
    // Placeholder enumeration until we round-trip on hardware. Most cheap
    // MPPTs of this generation expose at least these:
    Unknown    (0, "Unknown"),
    SealedLead (1, "Sealed lead-acid"),
    GelLead    (2, "Gel lead-acid"),
    FloodedLead(3, "Flooded lead-acid"),
    Lithium    (4, "Lithium (LiFePO4)");

    companion object {
        fun ofCode(c: Int): BatteryType = entries.firstOrNull { it.code == c } ?: Unknown
    }
}

@Serializable
enum class OutputMode(val code: Int, val displayName: String) {
    Manual(0, "Manual"),
    Auto  (1, "Auto / dusk-to-dawn"),
    Timer (2, "Timer"),
    Unknown(255, "Unknown");

    companion object {
        fun ofCode(c: Int): OutputMode = entries.firstOrNull { it.code == c } ?: Unknown
    }
}
